src/librustc_typeck/collect/type_of.rs

```diff
@@ -213,6 +234,45 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     .discr_type()
                     .to_ty(tcx),
 
+                Node::Expr(&Expr { kind: ExprKind::MethodCall(segment, span, _), .. }) => {
+                    // We want to know the type of `arg` in `x.foo::<{ arg }>()`.
+                    //
+                    // This means that we have to get the type of `x` first,
+                    // which requires us to first typeck the context which contains `x`.
+                    // As this might cause cycle errors, we use `type_of_const_arg` for generic arguments
+                    // during `typeck_tables_of` of the surrounding body.
+                    //
+                    // We can therefore freely use `generics_of(parenent_) (hopefully)
+                    let body_owner = tcx.hir().get_parent_did(parent_node_id);
+                    let tables = tcx.typeck_tables_of(body_owner.to_def_id());
```

Comment:
If you want to accelerate this further, use `typeck_tables_of` for all paths found in expressions ;).

The resulting cycle errors should guide you to finding all of the queries that need adapting.

Reply:
Done + done